File: version_3/code-samples/Indexes/DynamicFields.cs
using System.Collections.Generic;
using System.Linq;

using Raven.Client.Document;
using Raven.Client.Indexes;

namespace Raven.Documentation.CodeSamples.Indexes
{
	public class DynamicFields
	{
		#region dynamic_fields_1
		public class Product
		{
			public string Id { get; set; }
			public List<Attribute> Attributes { get; set; }
		}

		public class Attribute
		{
			public string Name { get; set; }
			public string Value { get; set; }
		}
		#endregion

		#region dynamic_fields_2
		public class Product_ByAttribute : AbstractIndexCreationTask<Product>
		{
			public Product_ByAttribute()
			{
				Map = products => from p in products
								  select new
								  {
									  _ = p.Attributes
										 .Select(attribute =>
											 CreateField(attribute.Name, attribute.Value, false, true))
								  };
			}
		}

		#endregion

		public void Dynamic()
		{
			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region dynamic_fields_3
					var products = session.Advanced.LuceneQuery<Product>("Product/ByAttribute")
						.WhereEquals("Color", "Red")
						.ToList();

					#endregion
				}
			}
		}
	}
}

File: Documentation/5.1/Samples/java/src/test/java/net/ravendb/Indexes/DynamicFields.java
package net.ravendb.Indexes;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.indexes.AbstractIndexCreationTask;
import net.ravendb.client.documents.session.IDocumentSession;

import java.util.List;

public class DynamicFields {

    //region dynamic_fields_1

    public static class Product {
        private String id;
        private List<Attribute> attributes;

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public List<Attribute> getAttributes() {
            return attributes;
        }

        public void setAttributes(List<Attribute> attributes) {
            this.attributes = attributes;
        }
    }

    public static class Attribute {
        private String name;
        private String value;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getValue() {
            return value;
        }

        public void setValue(String value) {
            this.value = value;
        }
    }
    //endregion

    //region dynamic_fields_2
    public static class Products_ByAttribute extends AbstractIndexCreationTask {
        public Products_ByAttribute() {
            map = "docs.Products.Select(p => new { " +
                "    _ = p.Attributes.Select(attribute => this.CreateField(attribute.name, attribute.value, false, true)) " +
                "})";
        }
    }
    //endregion

    public DynamicFields() {
        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                //region dynamic_fields_4
                List<Product> results = session
                    .query(Product.class, Products_ByAttribute.class)
                    .whereEquals("color", "red")
                    .toList();
                //endregion
            }
        }
    }

    //region syntax
    object CreateField(string name, object value);

    object CreateField(string name, object value, bool stored, bool analyzed);

    object CreateField(string name, object value, CreateFieldOptions options);
    //endregion
}
